extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3Utils.java

```diff
@@ -88,121 +84,172 @@ public boolean apply(Throwable e)
   }
 
-  public static String toFilename(String key)
+  static String toFilename(String key)
   {
     return toFilename(key, "");
   }
 
-  public static String toFilename(String key, final String suffix)
+  static String toFilename(String key, final String suffix)
   {
     String filename = key.substring(key.lastIndexOf("/") + 1); // characters after last '/'
     filename = filename.substring(0, filename.length() - suffix.length()); // remove the suffix from the end
     return filename;
   }
+
+  static AccessControlList grantFullControlToBucketOwver(AmazonS3Client s3Client, String bucket)
+  {
+    final AccessControlList acl = s3Client.getBucketAcl(bucket);
+    acl.grantAllPermissions(new Grant(new CanonicalGrantee(acl.getOwner().getId()), Permission.FullControl));
+    return acl;
+  }
+
+  public static String extractS3Key(URI uri)
+  {
+    return uri.getPath().startsWith("/") ? uri.getPath().substring(1) : uri.getPath();
+  }
+
+  // TODO: check this is possible with the official sdk
+  // Copied from org.jets3t.service.model.StorageObject.isDirectoryPlaceholder()
+  public static boolean isDirectoryPlaceholder(String key, ObjectMetadata objectMetadata)
+  {
+    // Recognize "standard" directory place-holder indications used by
+    // Amazon's AWS Console and Panic's Transmit.
+    if (key.endsWith("/") && objectMetadata.getContentLength() == 0) {
+      return true;
+    }
+    // Recognize s3sync.rb directory placeholders by MD5/ETag value.
+    if ("d66759af42f282e1ba19144df2d405d0".equals(objectMetadata.getETag())) {
+      return true;
+    }
+    // Recognize place-holder objects created by the Google Storage console
+    // or S3 Organizer Firefox extension.
+    if (key.endsWith("_$folder$") && objectMetadata.getContentLength() == 0) {
+      return true;
+    }
+
+    // We don't use JetS3t APIs anymore, but the below check is still needed for backward compatibility.
+
+    // Recognize legacy JetS3t directory place-holder objects, only gives
+    // accurate results if an object's metadata is populated.
+    if (objectMetadata.getContentLength() == 0 && MIMETYPE_JETS3T_DIRECTORY.equals(objectMetadata.getContentType())) {
+      return true;
+    }
+    return false;
+  }
+
+  public static S3ObjectSummary getSingleObjectSummary(AmazonS3Client s3Client, String bucket, String key)
+  {
+    final ListObjectsV2Request request = new ListObjectsV2Request()
+        .withBucketName(bucket)
+        .withPrefix(key)
+        .withMaxKeys(1);
+    final ListObjectsV2Result result = s3Client.listObjectsV2(request);
+
+    if (result.getKeyCount() == 0) {
+      throw new ISE("Cannot find object for bucket[%s] and key[%s]", bucket, key);
+    }
+    final S3ObjectSummary objectSummary = result.getObjectSummaries().get(0);
+    if (!objectSummary.getBucketName().equals(bucket) || !objectSummary.getKey().equals(key)) {
+      throw new ISE("Wrong object[%s] for bucket[%s] and key[%s]", objectSummary, bucket, key);
```

Comment:
Does this mean that if there are multiple keys that match the prefix, then if the returned value isn't the first one in the result this will fail?  that seems kind of fragile.

Reply:
Yes, I added javadoc for this method.